lib/rspec/core/memoized_helpers.rb

```diff
@@ -123,14 +123,28 @@ def is_expected
 
       # @private
       def __memoized
-        @__memoized ||= {}
+        @__memoized ||= Memoized.new
+      end
+
+      # @private
+      class Memoized
+        def initialize
+          @memoized = {}
+          @mutex = ReentrantMutex.new
+        end
+
+        def fetch_or_store(key)
+          @mutex.synchronize do
+            @memoized.fetch(key) { @memoized[key] = yield }
+          end
+        end
```

Comment:
That's an interesting idea. I'll benchmark them, b/c I'm unsure if the cost of two fetches is lower than the cost of the synchronize. Might depend on access patterns.